storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -1106,6 +1108,93 @@ public void setFinalizedDepositSnapshot(final DepositTreeSnapshot finalizedDepos
     }
   }
 
+  @Override
+  public Optional<UInt64> getFirstCustodyIncompleteSlot() {
+    return dao.getFirstCustodyIncompleteSlot();
+  }
+
+  @Override
+  public Optional<UInt64> getFirstSamplerIncompleteSlot() {
+    return dao.getFirstSamplerIncompleteSlot();
+  }
+
+  @Override
+  public Optional<DataColumnSidecar> getSidecar(final DataColumnSlotAndIdentifier identifier) {
+    final Optional<Bytes> maybePayload = dao.getSidecar(identifier);
+    return maybePayload.map(payload -> spec.deserializeSidecar(payload, identifier.slot()));
+  }
+
+  @Override
+  public Optional<DataColumnSidecar> getNonCanonicalSidecar(
+      final DataColumnSlotAndIdentifier identifier) {
+    final Optional<Bytes> maybePayload = dao.getNonCanonicalSidecar(identifier);
+    return maybePayload.map(payload -> spec.deserializeSidecar(payload, identifier.slot()));
+  }
+
+  @Override
+  @MustBeClosed
+  public Stream<DataColumnSlotAndIdentifier> streamDataColumnIdentifiers(
+      final UInt64 firstSlot, final UInt64 lastSlot) {
+    return dao.streamDataColumnIdentifiers(firstSlot, lastSlot);
+  }
+
+  @Override
+  @MustBeClosed
+  public Stream<DataColumnSlotAndIdentifier> streamNonCanonicalDataColumnIdentifiers(
+      final UInt64 firstSlot, final UInt64 lastSlot) {
+    return dao.streamNonCanonicalDataColumnIdentifiers(firstSlot, lastSlot);
+  }
+
+  @Override
+  public Optional<UInt64> getEarliestDataColumnSidecarSlot() {
+    return dao.getEarliestDataSidecarColumnSlot();
+  }
+
+  @Override
+  public void setFirstCustodyIncompleteSlot(final UInt64 slot) {
+    try (final FinalizedUpdater updater = finalizedUpdater()) {
+      updater.setFirstCustodyIncompleteSlot(slot);
+      updater.commit();
+    }
+  }
+
+  @Override
+  public void setFirstSamplerIncompleteSlot(final UInt64 slot) {
+    try (final FinalizedUpdater updater = finalizedUpdater()) {
+      updater.setFirstSamplerIncompleteSlot(slot);
+      updater.commit();
+    }
+  }
+
+  @Override
+  public void addSidecar(final DataColumnSidecar sidecar) {
+    try (final FinalizedUpdater updater = finalizedUpdater()) {
+      updater.addSidecar(sidecar);
+      updater.commit();
+    }
+  }
+
+  @Override
+  public void addNonCanonicalSidecar(final DataColumnSidecar sidecar) {
+    try (final FinalizedUpdater updater = finalizedUpdater()) {
+      updater.addNonCanonicalSidecar(sidecar);
+      updater.commit();
+    }
+  }
+
+  @Override
+  public void pruneAllSidecars(final UInt64 tillSlotInclusive) {
```

Comment:
ok so worst case this could be very time consuming and hold a TX open for ages. we should note this to come back to and fix at a minimum. the pattern is bad and it'll cause us a lot of pain.
It's the only use of `streamDataColumnIdentifiers` and the less streams we have from the DAO the better also, as a consideration, but at a very minimum it needs to batch TX's and not hold things open for ages.

Reply:
Created https://github.com/Consensys/teku/issues/9437 to track this.